doc/orml.1
.\"t
.\" Automatically generated by Pandoc 1.19.1
.\"
.TH "orml" "1" "" "April 13, 2017" "0.9"
.hy
.SH NAME
.PP
orml \- orml rules many lands.
.SH SYNOPSIS
.PP
orml [version] | [help] | [drop \f[I][entry]\f[]] | [list \f[I][\@hidden
| [pattern | []]]\f[]] | [move \f[I][from] [to]\f[]] | [install
\f[I][key]\f[]] | [insert \f[I][entry] [\- | input]\f[]] | [select
\f[I][entry]\f[]] | [hide \f[I][entry]\f[]] | [unhide \f[I][entry]\f[]]
| [export \f[I][file]\f[]] | [import \f[I][file]\f[]] | [\-\-passphrase
\f[I][passphrase]\f[]] | [\-\-secret \f[I][length]\f[]] | [\-\-as
\f[I][key]\f[]] | [\-\-encrypt] | [\-\-decrypt] | [\-\-clipboard] |
[\-\-password] | [\-\-null] | [\-\-force]
.SH DESCRIPTION
.PP
Think of it like pass, except with more features and a more maintainable
codebase.
orml is a password/data manager which allows for various additional
security and convenience features.
.SH DEPENDENCIES
.PP
.TS
tab(@);
r l l.
T{
Dependency
T}@T{
version
T}@T{
purpose
T}
_
T{
bash
T}@T{
>= 4.0.0
T}@T{
core
T}
T{
gpg
T}@T{
>= 2.0
T}@T{
core
T}
T{
gnu\-getopt
T}@T{
*
T}@T{
core
T}
T{
make
T}@T{
*
T}@T{
installation
T}
T{
git
T}@T{
*
T}@T{
installation
T}
T{
tar
T}@T{
*
T}@T{
command:import, command:export
T}
T{
tree
T}@T{
*
T}@T{
command:list
T}
T{
xclip#pbcopy
T}@T{
*
T}@T{
opts:clipboard
T}
.TE
.SH INSTALLING
.PP
Make is used for installing and removing orml.
The commands build & clean can be used for this, build being the
default.
.IP
.nf
\f[C]
$\ git\ clone\ https://github.com/muse/orml
$\ make\ build
\f[]
.fi
.PP
After that, you will need to install orml itself, before you do this
make sure you have a GPG key.
You can validate this with \f[C]$\ gpg\ \-\-list\-secret\-keys\f[] which
lists the secret keys available on the system, orml will use this to
populate ~/.orml/keys.
.IP
.nf
\f[C]
$\ orml\ install\ D16BCBA7
\f[]
.fi
.PP
Before entering any commands validate that ~/.orml/keys has your default
key at the top.
When the \f[I]\-\-as\f[] option is omited orml will use this instead.
.SH COMMANDS
.SS version
.PP
Display the current version.
.SS help
.PP
Open the manual page if available and possible.
.SS drop [entry]
.PP
Drop the matching entry, hidden entry or directory.
This will prompt a confirmation which can optionally be skipped with
\f[I]\-\-force\f[].
.SS install [key]
.PP
Refer to the INSTALLING section above.
.SS list [\@hidden | [pattern | []]]
.PP
List entries optionally based on the provided argument.
.IP \[bu] 2
When no arguments are specified, all of the entries will be list
.IP \[bu] 2
When a single argument matching \f[I]\@hidden\f[] is specified, only
~/.orml/.hidden will be listed
.IP \[bu] 2
Any other arguments will be used to pattern match entries, and list
those exclusively
.SS move [from] [to]
.PP
Move the first entry to the second location.
When the second location exists, a prompt will appear which can be
skipped with \f[I]\-\-force\f[].
.SS insert [entry] [\- | input]
.PP
Insert and encrypt a new file, text message or secure/private key.
The key used to encrypt this entry is taken from ~/.orml/keys by
default, it will take the first line in the file.
This is also the way for you to set a preference on which key to use.
You can specify a key which will overwrite the preference with
\f[I]\-\-as [key]\f[].
There are multiple ways to provide insert with data, this is done by
changing the second argument.
.IP \[bu] 2
Omitting it, which will spawn a prompt.
This prompt will not visibly hide user input without the
\f[I]\-\-password\f[] option
.IP \[bu] 2
Supplying a single dash, telling orml to read from standard input
instead of using the arguments
.IP \[bu] 2
Specifying the \f[I]\-\-secret [length]\f[] option, this will use a
/dev/urandom string using the printable ascii character set and the
provided length
.IP \[bu] 2
Anything else which doesn\[aq]t apply to the previous clauses is
initially treated as a file path.
If this ends up not being a valid path, it is interpret as text.
.SS select [entry]
.PP
Select and decrypt a existing entry.
You don\[aq]t need to specify a key, gpg will identify if the required
key is available by itself.
It is possible to suppress any output to standard output with the
\f[I]\-\-null\f[] option.
Additionally you can use the \f[I]\-\-clipboard\f[] option to copy the
contents of entry to your clipboard.
.SS hide [entry]
.PP
Hide the matching entry.
the file is then moved to ~/.orml/hidden with the hash of the relative
path as its new filename.
You can still access this entry like you normally would with select.
.SS unhide [entry]
.PP
Unhide the matching hidden entry by replacing the hash with the provided
matching relative path, it will be interpret the same way it was set
initially.
.SS import [file]
.PP
Import a tarball or gpg file resembling a logical ordering structure.
All found directories are created or reused dependent on the existing
directories in ~/.orml.
All the new files found are added, the duplicate files will be
overwritten by their import.
When a gpg file is provided \f[B]without\f[] extension, you need to tell
provide the \f[I]\-\-decrypt\f[] option.
.SS export [file]
.PP
Export everything in ~/.orml by making a tarball writen to the provided
\&.
You can additionally encrypt it with the \f[I]\-\-encrypt\f[] option,
this can be used in combination with the \f[I]\-\-as [key]\f[].
.SH OPTIONS
.SS \-\-passphrase [passphrase]
.PP
Use this passphrase instead of the later prompted one (not very secure).
.SS \-\-secret [length]
.PP
Ignore any user received input and use a /dev/urandom string consisting
of all the printable ascii characters and the provided length.
.SS \-\-as [key]
.PP
Do not use the preferenced key from ~/.orml/keys.
Instead use the provided key, regardless of its existence in the
~/.orml/keys file.
This option only applies to \f[B]export\f[] when the
\f[I]\-\-encrypt\f[] option is also set.
.SS \-\-encrypt
.PP
After creating the tarball, encrypt it with either the preferenced key,
or the key provided by \f[I]\-\-as\f[].
.SS \-\-decrypt
.PP
Force decrypt the tarball before attempting to import it.
.SS \-\-clipboard
.PP
Copy the contents of the selected entry to the clipboard.
.SS \-\-null
.PP
Don\[aq]t show anything.
.SS \-\-password
.PP
Don\[aq]t show what\[aq]s being typed into a prompt.
.SS \-\-force
.PP
Don\[aq]t prompt.
.SH BUGS
.IP \[bu] 2
mvdw at airmail dot cc
.IP \[bu] 2
github/muse/orml


orml.bash
#!/usr/bin/env bash
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

set -o pipefail

# TODO:
#   Don't hardcode the library path, changing the Makefile will break orml entirely.
source /usr/local/lib/orml/util.bash
source /usr/local/lib/orml/command.bash

ORML_VERSION="0.9"
ORML_NAME="$0"
ORML_COMMAND="${1:-list}"
ORML_LONG="as:,passphrase:,secret:,force,hidden,password,clipboard,null,encrypt,decrypt"
ORML_SHORT=":"
ORML_ALLOC="${ORML_ALLOC:-$HOME}"
ORML_STORE="${ORML_STORE:-$ORML_ALLOC/.orml}"
ORML_TEMPLATE=${ORML_TEMPLATE:-$ORML_STORE/template}
ORML_KEYS="${ORML_KEYS:-$ORML_STORE/keys}"
ORML_HIDDEN="${ORML_HIDDEN:-$ORML_STORE/.hidden}"

ORML_OPTS_AS="$(head -1 "$ORML_KEYS" 2> /dev/null)"
ORML_OPTS_SECRET=-1
ORML_OPTS_PASSPHRASE=
ORML_OPTS_NULL=false
ORML_OPTS_ENCRYPT=false
ORML_OPTS_DECRYPT=false
ORML_OPTS_CLIPBOARD=false
ORML_OPTS_HIDDEN=false

export GPG_TTY=
GPG_TTY=$(tty 2> /dev/null)

function _argv {
    eval set -- "$(getopt \
                -o "$ORML_SHORT" \
                -l "$ORML_LONG" \
                -n "$ORML_NAME" \
                -- "$@")"
    while :; do
        case "$1" in
            --as)
                ORML_OPTS_AS="$2"
                shift 2 ;;
            --secret)
                ORML_OPTS_SECRET="$2"
                shift 2 ;;
            --passphrase)
                ORML_OPTS_PASSPHRASE="$2"
                shift 2 ;;
            --encrypt)
                ORML_OPTS_ENCRYPT=true
                shift 1 ;;
            --decrypt)
                ORML_OPTS_DECRYPT=true
                shift 1 ;;
            --hidden)
                ORML_OPTS_HIDDEN=true
                shift 1 ;;
            --password)
                ORML_OPTS_PASSWORD=true
                shift 1 ;;
            --force)
                ORML_OPTS_FORCE=true
                shift 1 ;;
            --clipboard)
                ORML_OPTS_CLIPBOARD=true
                shift 1 ;;
            --null)
                ORML_OPTS_NULL=true
                shift 1 ;;
            --)
                # We shift twice here because we save the command at the top of
                # the script, this will remove the command together with --.
                shift 2
                break ;;
            *)
                echo >&2 "Here be dragons."
                exit 1 ;;
        esac
    done
    for ARG; do
        ORML_ARGV+=("$ARG")
    done
}

# TODO:
#   Write error messages when not enough arguments are found.
_argv "$@"
case "$ORML_COMMAND" in
    version|help|install|list)
        (( ${#ORML_ARGV[@]} >= 0 )) || exit 1 ;;
    select|insert|export|hide|unhide|import|drop)
        (( ${#ORML_ARGV[@]} >= 1 )) || exit 1 ;;
    move)
        (( ${#ORML_ARGV[@]} >= 2 )) || exit 1 ;;
    *)
        echo >&2 "$ORML_COMMAND isn't a valid command"
        exit 1 ;;
esac
"_$ORML_COMMAND" "${ORML_ARGV[@]}" <&0
